src/vegasflow/vflowplus.py

```diff
@@ -0,0 +1,234 @@
+
+FBINS = float_me(BINS_MAX)
+
+
+@tf.function(input_signature=3 * [tf.TensorSpec(shape=[None, None], dtype=DTYPE)])
+def _compute_x(x_ini, xn, xdelta):
+    """ Helper function for ``generate_samples_in_hypercubes`` """
+    aux_rand = xn - tf.math.floor(xn)
+    return x_ini + xdelta * aux_rand
+
+
+# same as vegasflow generate_random_array
+@tf.function(input_signature=2*[tf.TensorSpec(shape=[None, None], dtype=DTYPE)])
+def _digest(xn, divisions):
+    ind_i = tf.cast(xn, DTYPEINT)
+    # Get the value of the left and right sides of the bins
+    ind_f = ind_i + ione
+    x_ini = tf.gather(divisions, ind_i, batch_dims=1)
+    x_fin = tf.gather(divisions, ind_f, batch_dims=1)
+    # Compute the width of the bins
+    xdelta = x_fin - x_ini
+    return ind_i, x_ini, xdelta
+
+
+@tf.function(
+    input_signature=[
+        tf.TensorSpec(shape=[None, None], dtype=DTYPE),
+        tf.TensorSpec(shape=[], dtype=DTYPEINT),
+        tf.TensorSpec(shape=[None], dtype=DTYPEINT),
+        tf.TensorSpec(shape=[None, None], dtype=DTYPEINT),
+        tf.TensorSpec(shape=[None, None], dtype=DTYPE),
+    ]
+)
+def generate_samples_in_hypercubes(rnds, n_strat, n_ev, hypercubes, divisions):
+    """Receives an array of random numbers 0 and 1 and
+    distribute them in each hypercube according to the
+    number of samples in each hypercube specified by n_ev
+
+    Parameters
+    ----------
+        `rnds`: tensor of random number between 0 and 1
+        `n_strat`: tensor with number of stratifications in each dimension
+        `n_ev`: tensor containing number of sample per hypercube
+        `hypercubes`: tensor containing all different hypercube
+        `divisions`: vegas grid
+
+    Returns
+    -------
+        `x` : random numbers collocated in hypercubes
+        `w` : weight of each event
+        `ind`: division index in which each (n_dim) set of random numbers fall
+        `segm` : segmentantion for later computations
+    """
+
+    indices = tf.repeat(tf.range(tf.shape(hypercubes, out_type=DTYPEINT)[0]), n_ev)
+    points = float_me(tf.gather(hypercubes, indices))
+    n_evs = float_me(tf.gather(n_ev, indices))
+    xn = tf.transpose((points + tf.transpose(rnds)) * FBINS / float_me(n_strat))
+    segm = indices
+
+    ind_i, x_ini, xdelta = _digest(xn, divisions)
+    # Compute the random number between 0 and 1
+    # This is the heavy part of the calc
+
+    x = _compute_x(x_ini, xn, xdelta)
+    # Compute the random number between the limits
+    #     x = reg_i + rand_x * (reg_f - reg_i)
+    # and the weight
+    weights = tf.reduce_prod(xdelta * FBINS, axis=0)
+    final_weights = weights / n_evs
+    x_t = tf.transpose(x)
+    int_xn = tf.transpose(ind_i)
+
+    return x_t, int_xn, final_weights, segm
+
+
+class VegasFlowPlus(VegasFlow):
+    """
+    Implementation of the VEGAS+ algorithm
+    """
+
+    def __init__(self, n_dim, n_events, train=True, adaptive=None, **kwargs):
+        super().__init__(n_dim, n_events, train, **kwargs)
```

Comment:
```suggestion
        _ = kwargs.setdefault("events_limit", n_events)
        super().__init__(n_dim, n_events, train, **kwargs)
```

Right now the integrator will try to break the integration into multiple devices which is not supported at this point by Vegas+ so in the meantime it should try to put all events in the same devices. This will ensure that it will work in CPU and in any GPU which enough memory to hold all data.

Reply:
Perfect, thanks!
I have a few ideas on how to modify it in order to break the integration.
I'll try something and I will let you know.